kythe/cxx/indexer/cxx/IndexerASTHooks.cc

```diff
@@ -5697,4 +5705,55 @@ void IndexerASTVisitor::LogErrorWithASTDump(absl::string_view msg,
   LOG(ERROR) << msg << " :" << std::endl << s;
 }
 
+void IndexerASTVisitor::PrepareAlternateSemanticCache() {
+  if (!DataflowEdges) return;
+  for (const auto& meta : Observer.meta()) {
+    for (const auto& rule : meta.second->rules()) {
+      GraphObserver::UseKind kind = GraphObserver::UseKind::kUnknown;
+      if (rule.second.semantic == MetadataFile::Semantic::kNone) continue;
+      switch (rule.second.semantic) {
+        case MetadataFile::Semantic::kWrite:
+          kind = GraphObserver::UseKind::kWrite;
+          break;
+        case MetadataFile::Semantic::kReadWrite:
+          kind = GraphObserver::UseKind::kReadWrite;
+          break;
+        default:
+          break;
+      }
+      auto end = Observer.getSourceManager()->getComposedLoc(meta.first,
+                                                             rule.second.end);
+      AlternateSemantic semantic{
+          end, kind, Observer.MintNodeIdForVName(rule.second.vname)};
+      auto loc = Observer.getSourceManager()->getComposedLoc(meta.first,
+                                                             rule.second.begin);
+      alternate_semantic_cache_[loc.getRawEncoding()] = semantic;
+    }
+  }
+}
+
+kythe::IndexerASTVisitor::AlternateSemantic*
+IndexerASTVisitor::AlternateSemanticForDecl(const clang::Decl* decl) {
+  if (!DataflowEdges) return nullptr;
+  const auto* canonical = decl->getCanonicalDecl();
+  if (auto found = alternate_semantics_.find(canonical);
+      found != alternate_semantics_.end()) {
+    return found->second;
+  }
+  const auto* named = llvm::dyn_cast<clang::NamedDecl>(decl);
+  // No sense in annotating an unnamed declaration. We still deal in `Decl`s
+  // in this function, though, because bare `Decl`s appear in many contexts
+  // (like the callee of a call expression).
+  if (named == nullptr) return nullptr;
+  auto range = RangeForNameOfDeclaration(named);
+  if (auto cached =
+          alternate_semantic_cache_.find(range.getBegin().getRawEncoding());
+      cached != alternate_semantic_cache_.end() &&
+      cached->second.end == range.getEnd()) {
```

Comment:
It looks like you're co-mingling the key and the value here, rather than just using a pair for the key?